File: backends/java/zappy-application/src/main/java/com/zappymart/application/accounts/LogInCustomer.java
package com.zappymart.application.accounts;

import com.zappymart.application.Visitor;
import com.zappymart.application.ports.CustomerRepository;
import com.zappymart.application.ports.PasswordHasher;
import com.zappymart.application.ports.RateLimiter;
import com.zappymart.application.ports.UnitOfWork;
import com.zappymart.domain.accounts.Customer;
import com.zappymart.domain.shared.EmailAddress;
import com.zappymart.domain.shared.Result;
import com.zappymart.domain.shared.UserErrorCode;

import java.util.Optional;

public final class LogInCustomer {

    private final UnitOfWork unitOfWork;
    private final CustomerRepository customerRepository;
    private final PasswordHasher passwordHasher;
    private final SessionOpening sessionOpening;
    private final AnonymousHandover anonymousHandover;
    private final RateLimiter rateLimiter;
    private final String hashThatMatchesNothing;

    public LogInCustomer(UnitOfWork unitOfWork, CustomerRepository customerRepository,
                         PasswordHasher passwordHasher, SessionOpening sessionOpening,
                         AnonymousHandover anonymousHandover, RateLimiter rateLimiter) {
        this.unitOfWork = unitOfWork;
        this.customerRepository = customerRepository;
        this.passwordHasher = passwordHasher;
        this.sessionOpening = sessionOpening;
        this.anonymousHandover = anonymousHandover;
        this.rateLimiter = rateLimiter;
        this.hashThatMatchesNothing = passwordHasher.hash("no customer has this password");
    }

    public Result<Authentication> execute(Visitor visitor, String typedEmail, String password, String device) {
        Optional<EmailAddress> email = EmailAddress.parse(typedEmail);
        if (email.isEmpty()) {
            return credentialsInvalid();
        }
        if (!rateLimiter.allows("login:" + email.get().value())) {
            return Result.refuse(UserErrorCode.RATE_LIMITED,
                    "Too many attempts in a short time. Wait a moment and try again.");
        }
        return unitOfWork.inTransaction(() -> {
            Optional<Customer> customer = customerRepository.byEmail(email.get());
            String storedHash = customer.map(Customer::passwordHash).orElse(hashThatMatchesNothing);
            boolean passwordFits = passwordHasher.matches(password, storedHash);
            if (customer.isEmpty() || !passwordFits) {
                return credentialsInvalid();
            }
            anonymousHandover.toCustomer(visitor, customer.get().id());
            return Result.of(sessionOpening.openFor(customer.get(),
                    device == null ? visitor.device() : device));
        });
    }

    private Result<Authentication> credentialsInvalid() {
        return Result.refuse(UserErrorCode.CREDENTIALS_INVALID,
                "That email address and password do not go together.");
    }
}

File: backends/kotlin/zappy-application/src/main/kotlin/nl/zappymart/application/accounts/LogInCustomer.kt
package nl.zappymart.application.accounts

import nl.zappymart.application.Visitor
import nl.zappymart.application.cart.VisitorCart
import nl.zappymart.application.ports.CustomerRepository
import nl.zappymart.application.ports.PasswordHasher
import nl.zappymart.application.ports.RateLimiter
import nl.zappymart.application.ports.UnitOfWork
import nl.zappymart.domain.accounts.PasswordHash
import nl.zappymart.domain.shared.EmailAddress
import nl.zappymart.domain.shared.Result
import nl.zappymart.domain.shared.UserErrorCode
import nl.zappymart.domain.shared.refusal

data class LoginRequest(val email: String, val password: String, val device: String)

class LogInCustomer(
    private val customers: CustomerRepository,
    private val passwords: PasswordHasher,
    private val signIn: SignIn,
    private val visitorCart: VisitorCart,
    private val wishlistOwner: WishlistOwner,
    private val rateLimiter: RateLimiter,
    private val unitOfWork: UnitOfWork,
) {

    private val hashThatMatchesNobody: PasswordHash by lazy {
        passwords.hash("a password that belongs to no customer at all")
    }

    fun execute(visitor: Visitor, request: LoginRequest): Result<Authentication> {
        val address = request.email.trim().lowercase()
        if (!rateLimiter.allows("login:$address")) {
            return refusal(UserErrorCode.RATE_LIMITED, "Too many attempts. Please wait and try again.")
        }
        return unitOfWork.execute {
            val customer = EmailAddress.of(address).let { parsed ->
                if (parsed is Result.Success) customers.findByEmail(parsed.value) else null
            }
            val hash = customer?.passwordHash ?: hashThatMatchesNobody
            val matches = passwords.matches(request.password, hash)
            if (customer == null || !matches) {
                refusal(
                    UserErrorCode.CREDENTIALS_INVALID,
                    "That email address and password do not match a customer.",
                )
            } else {
                wishlistOwner.moveToCustomer(visitor, customer.id)
                visitorCart.moveToCustomer(visitor, customer.id)
                Result.Success(signIn.start(customer, request.device))
            }
        }
    }
}
